README.rst
Sphere Base
############

Introduction
=============

This library was created as a building block for use in applications were information should be displayed as nodes or
labels on a sphere. These nodes are interactive so they can be moved by dragging and can be connected to other nodes.


* Development Status: 4 - Beta
* Free software: MIT license


Features
=========

- Customizable spheres with textures and colors
- Nodes, sockets and edges can be dragged over the surface of the sphere
- Support for undo/redo, cut and paste using serialization
- Hovering effects, dragging nodes, cutting edges
- Background skybox image can be changed
- Sphere color and texture can be changed
- Example on how the library can be implemented


Requirements
=============

This package is created in python using a pyqt5 window.

- Python 3.x
- PyOpenGL
- PyQt5
- pyperclip
- pybullet
- Pillow
- numpy


Supported Environment
======================

* Windows (Win32, x64)


The library is available from PyPI

    $ pip install sphere-base





Or download the source code from github

    git clone https://github.com/rboltze/sphere_base.git








HISTORY.rst
=======
History
=======

0.1.0 (2023-04-16) Beta
------------------

* Updated mouse pointers. All functionality is now in the library. Releasing as Beta version

0.0.8 (2023-04-14) Alfa
------------------

* Updated dragging the sphere rotating around the y-axis

0.0.7 (2023-04-12) Alfa
------------------

* Updated menu system

0.0.6 (2023-04-8) Alfa
------------------

* Undo, redo implemented

0.0.5 (2023-03-29) Alfa
------------------

* Edges are now dynamically created and stored in Opengl VBO's

0.0.4 (2023-03-29) Alfa
------------------

* Added copying and pasting using clipboard serialisation

0.0.3 (2023-03-28) Alfa
------------------

* Updated positioning over the target sphere using mouse_ray_collision_points

0.0.2 (2023-03-21) Alfa
------------------

* Added a settings screen to the example to change sphere colors and texture

0.0.1 (2023-03-15) Alfa
------------------

* First release as a library


sphere_base/constants.py
# -*- coding: utf-8 -*-

"""
Constants used in Sphere_IOT project.

These constants include camera properties for the OpenGL shaders. Dictionary with shader data and lists with textures
and resources data.

"""
SPHERE_SMALL_RADIUS = 0.01
SPHERE_RADIUS = 3
NODE_DISC_RADIUS = 0.075
SOCKET_RADIUS = 0.015
HOVER_MIN_DISTANCE = 10

WIDTH, HEIGHT = 720, 720

# camera props for shader
FOV = 45.0
NEAR_VAL = 0.1
FAR_VAL = 100

# camera props
CAMERA_UP = 0
ROTATION = 90
MAX_YAW_UP = 70
MIN_YAW_DOWN = -70
CAM_MOVEMENT_STEPS = 4

MESH_DIR = "..//sphere_base/model/resources/meshes/"
TEXTURE_DIR = "..//sphere_base/model/resources/"
TEXTURES_DIR = "..//sphere_base/model/resources/textures/"
ICONS_DIR = "..//sphere_base/model/resources/icons/"
SPHERE_TEXTURE_DIR = "..//sphere_base/model/resources/sphere_textures/"
SHADER_DIR = "..//sphere_base/model/resources/shaders/"
SKYBOX_IMG_DIR = "..//sphere_base/model/resources/textures/skybox/"

MODELS = {
    "sphere_base": {"model_id": 1, "model_file_name": "sphere1.obj", "shader": "SphereShader",
                    "vertex_shader": "vert_sphere.glsl",
                    "fragment_shader": "frag_sphere.glsl", "geometry_shader": "none"},
    "holo_sphere": {"model_id": 2, "model_file_name": "sphere1.obj", "shader": "HoloSphereShader",
                    "vertex_shader": "vert_holo_sphere.glsl", "fragment_shader": "frag_holo_sphere.glsl",
                    "geometry_shader": "none"},
    "sphere_small": {"model_id": 3, "model_file_name": "sphere_small.obj", "shader": "SphereSmallShader",
                     "vertex_shader": "vert_sphere_small.glsl", "fragment_shader": "frag_sphere_small.glsl",
                     "geometry_shader": "none"},
    "skybox": {"model_id": 4, "model_file_name": "skybox.obj", "shader": "SkyboxShader",
               "vertex_shader": "vert_skybox.glsl",
               "fragment_shader": "frag_skybox.glsl", "geometry_shader": "none"},
    "node": {"model_id": 5, "model_file_name": "node_disc1.obj", "shader": "NodeShader",
             "vertex_shader": "vert_node.glsl",
             "fragment_shader": "frag_node.glsl", "geometry_shader": "none"},
    "socket": {"model_id": 6, "model_file_name": "node_disc1.obj", "shader": "SocketShader",
               "vertex_shader": "vert_socket.glsl",
               "fragment_shader": "frag_socket.glsl", "geometry_shader": "none"},
    "rubber_band": {"model_id": 7, "model_file_name": "square1x1.obj", "shader": "FlatShader",
                    "vertex_shader": "vert_2d.glsl",
                    "fragment_shader": "frag_2d.glsl", "geometry_shader": "none"},
    "circle": {"model_id": 8, "model_file_name": "vertex1.obj", "shader": "CircleShader",
               "vertex_shader": "vert_circle.glsl",
               "fragment_shader": "frag_circle.glsl", "geometry_shader": "geom_circle.glsl"},
    "square": {"model_id": 9, "model_file_name": "vertex1.obj", "shader": "SquareShader",
               "vertex_shader": "vert_square.glsl",
               "fragment_shader": "frag_square.glsl", "geometry_shader": "geom_square.glsl"},
    "edge": {"model_id": 10, "model_file_name": "line_1x1.obj", "shader": "EdgeShader",
              "vertex_shader": "vert_sphere_edge.glsl",
              "fragment_shader": "frag_sphere_edge.glsl", "geometry_shader": "none"},
    "cross_hair1": {"model_id": 11, "model_file_name": "vertex1.obj", "shader": "CrossShader",
                    "vertex_shader": "vert_circle.glsl",
                    "fragment_shader": "frag_circle.glsl", "geometry_shader": "geom_cross.glsl"},
    "cube": {"model_id": 12, "model_file_name": "vertex1.obj", "shader": "SphereShader",
             "vertex_shader": "vert_sphere.glsl",
             "fragment_shader": "frag_sphere.glsl", "geometry_shader": "none"},
    "cube_sphere": {"model_id": 13, "model_file_name": "cubesphere.obj", "shader": "SphereShader",
                    "vertex_shader": "vert_sphere.glsl",
                    "fragment_shader": "frag_sphere.glsl", "geometry_shader": "none"},
    "drag_edge": {"model_id": 14, "model_file_name": "sphere1.obj", "shader": "DragEdgeShader",
                  "vertex_shader": "vert_sphere.glsl",
                  "fragment_shader": "frag_sphere.glsl", "geometry_shader": "none"},
    "sphere_lines": {"model_id": 15, "model_file_name": "line_1x1.obj", "shader": "EdgeShader",
                     "vertex_shader": "vert_sphere_edge.glsl",
                     "fragment_shader": "frag_sphere_edge.glsl", "geometry_shader": "none"},
}

SHADER_SWITCH = {"node": 0, "sphere_base": 1, "square1x1": 2, "rubber_band": 2, "sphere_small":
                 1, "socket": 2, "cube": 1, "cube_sphere": 1}

EDGE_TYPE_DIRECT = 1
TRANSPARENCY_SMALL_SPHERES = 0.8
TRANSPARENCY_DETAIL_SPHERE = 0.5
